### PR TITLE
Feat/search timeranges

### PR DESCRIPTION
**Pull request by tarnung**
First draft of functionality. Discussion is over at #504
There is no generic 'date:' search yet.
There is no 'desc:' text search. Shouldn't be too difficult but it should probably be its own pr.

### CODE REVIEW

**Comment by munen** [comment]
Sweet, you're moving very fast! :+1: 

If it helps you, the headline filter has lots of unit tests:

https://github.com/200ok-ch/organice/blob/106f8cc27e9bd623689417e77b037e387041a19d/src/lib/headline_filter.unit.test.js

There's also the option to add integration tests like this one to ensure that the feature works as expected from the search component:

https://github.com/200ok-ch/organice/blob/106f8cc27e9bd623689417e77b037e387041a19d/src/components/OrgFile/OrgFile.integration.test.js#L323-L341

File: src/lib/headline_filter.js
Comment (schoettl) [inline src/lib/headline_filter.js:202]: Sorry, I didn't fully analyze your new code. Just a question: Have you considered that filter terms can be negated with "-"? I don't remember if that applies for all filter terms or just for the existing ones.

Does it make sense for date queries? Not necessarily, I think. But in that case, maybe the search grammar must be changed.

**Review by schoettl — CHANGES_REQUESTED** [review]
Wow, amazing work so far. I haven't tried and did not fully review. Just added a few remarks...

File: src/lib/headline_filter.js
Comment (tarnung) [reply src/lib/headline_filter.js:202]: Something like `clock:2020 -clock:202004` meaning all year without april makes sense. I don't know if it is that useful though.
Seems like I just ignore negated timerange-searchqueries at the moment. That's not expected behaviour. Since for now all filters are negateable it would make sense to do it here too.

File: src/lib/headline_filter_parser.grammar.pegjs
Comment (tarnung) [reply src/lib/headline_filter_parser.grammar.pegjs:118]: Let's discuss precision further over at #504. I don't think I'd often use a more granular search, but I guess it would not hurt to allow for it.

File: src/lib/headline_filter.js
Comment (schoettl) [reply src/lib/headline_filter.js:202]: But don't start too complex ;) you can as well just throw a "parse exception" or something to tell the user that negating this filter is not yet supported. Implement the important (and practical) features first.

File: src/lib/headline_filter_parser.grammar.pegjs
Comment (tarnung) [reply src/lib/headline_filter_parser.grammar.pegjs:103]: it does not seem to work

File: src/lib/headline_filter_parser.grammar.pegjs
Comment (tarnung) [reply src/lib/headline_filter_parser.grammar.pegjs:124]: I guess you are right, but I wanted to be consistent with TimeOffset, TimeUnit etc.

File: src/lib/headline_filter.js
Comment (tarnung) [reply src/lib/headline_filter.js:203]: I throw. 

I'm not quite sure if the logic should be the same as in text search.
In text search if a header text matches a filter and an exclusion filter, it is not shown.
If a header has a timestamp that matches a filter and an exclusion filter it should also not be shown.
But what if a header has one timestamp that matches a filter and another timestamp that matches an exclusion filter? I think it should be included but it's not clear cut.
=> negated timeranges would be easy if we only allowed for one time filter in a search string. Since there can be an arbitrary amount of filters, they are tricky. I won't touch them for now,

File: src/lib/headline_filter_parser.grammar.pegjs
Comment (schoettl) [reply src/lib/headline_filter_parser.grammar.pegjs:124]: That's what I thought. But inventing new spellings for words just to be consistent with other words is not good practice IMO. TimeTimestamp would be consistent and correct ^^

File: src/lib/headline_filter.js
Comment (schoettl) [reply src/lib/headline_filter.js:203]: That even a human can't easily answer :D My feeling is that exclusion should have "the last word", i.e. exclude headline if an exclude filter matches.

But not allowing these things is good as well. It also makes the whole thing easier to understand.

File: src/lib/headline_filter_parser.grammar.pegjs
Comment (tarnung) [reply src/lib/headline_filter_parser.grammar.pegjs:106]: Rejoice, I added even more duplication to get things to work ;)

File: src/lib/headline_filter_parser.grammar.pegjs
Comment (tarnung) [reply src/lib/headline_filter_parser.grammar.pegjs]: I surrendered and renamed it Timestamp

**Comment by tarnung** [comment]
Points for discussion:
- I limited scheduled and deadline to only search for active timestamps. Searching with the date keyword also includes inactive timestamps. Is that ok or should it be limited too?
- When repeating tasks, organice writes an active timestamp into the logbook to record the status change. That timestamp is not parsed and appears as text. I guess it should be an inactive timestamp. I don't know if you plan on parsing them. For now they are ignored.
- I defer date checking to `new Date('2020-02-31')`. Firefox marks this as an illegal date. Chrome accepts days up to 31 for all months and maps them onto the first days of the following month. There is a way to pass the date-fn `isExists` into the parser via the options variable. But then we could no longer use the online version on the pegjs website which is quite handy. 
- When I began to write tests, all the existing tests passed. Soon thereafter some tests randomly started to fail. 
- The pipeline is not happy with my tests. They worked fine for me though.

**Comment by munen** [comment]
> - I limited scheduled and deadline to only search for active
>   timestamps. Searching with the date keyword also includes inactive
>   timestamps. Is that ok or should it be limited too?

I would make them behave the same. My original proposal was:

> Hence, I propose a date: qualifier that would work across all timestamps. That would also be in line with how the Emacs Org agenda works (limiting to a day or week will take into account SCHEDULED, DEADLINE and plain active timestamps).

schoettl gave a 👍 on that.

My reason is that "an inactive timestamp that does not cause an agenda entry" (https://orgmode.org/manual/Creating-Timestamps.html#Creating-Timestamps).

Do you have a reason that you want `date` to include inactive timestamps? If we should have a search on inactive timestamps, I propose to add a dedicated flag for that like `inactive: true`.

> - When repeating tasks, organice writes an active timestamp into the
>   logbook to record the status change. That timestamp is not parsed
>   and appears as text. I guess it should be an inactive timestamp. I
>   don't know if you plan on parsing them. For now they are ignored
>   by search.

That's a bug. Good catch! I've created an issue for that: https://github.com/200ok-ch/organice/issues/512

> - I defer date checking to `new Date('2020-02-31')`. Firefox marks
>  this as an illegal date. Chrome accepts days up to 31 for all
>  months and maps them onto the first days of the following month.
>  There is a way to pass the date-fn `isExists` into the parser via
>  the options variable. But then we could no longer use the online
>  version on the pegjs website which is quite handy. For now the
>  behaviour depends on the browser you use.

Maybe @schoettl can comment on that. I haven't worked with pegjs before.

NB: Emacs auto-corrects a timestamp like [2020-02-31] into [2020-03-02] which I only just now learned about. Quite handy(;

> - When I began to write tests, all the existing tests passed. Soon
>   thereafter some tests randomly started to fail. They had nothing
>   to do with what i did so I ignored them.

That doesn't sound alright. Can you point me to the tests that randomly started to fail? So far, one user reported that one test in particular was failing randomly. Personally I have seen only one test fail randomly once (on CI). Here's the issue we have created for it: https://github.com/200ok-ch/organice/issues/442

I haven't seen this behavior since. But if you're experiencing it, too, then it would certainly be good to track it down.


> - The pipeline is not happy with my tests. They worked fine for me
>  though.

There are two kinds of errors that you probably didn't check, because you 'only' ran the tests:

1. eslint warnings that show up on `yarn start`. I have fixed them here: https://github.com/200ok-ch/organice/pull/505/commits/e833194f84fa90bf489971704c553a89897263d5
2. eslint errors that show up on `yarn eslint`. I have fixed them here: https://github.com/200ok-ch/organice/pull/505/commits/c80fe729bdf0ca15f77f4807ea61f19a2308bc3e

CI is now happy, too(;

**Comment by schoettl** [comment]
>     * I defer date checking to `new Date('2020-02-31')`. Firefox marks this as an illegal date. Chrome accepts days up to 31 for all months and maps them onto the first days of the following month. There is a way to pass the date-fn `isExists` into the parser via the options variable. But then we could no longer use the online version on the pegjs website which is quite handy.  For now the behaviour depends on the browser you use.

I haven't used the online pegjs parser but if you say it's handy, I would try to not introduce dependencies like date-fn.

The difference in the browser is IMO a minor problem which can be ignored (and not our fault ;) ). As long as the app doesn't crash but shows a error message or red border around the search input I'm fine :)

**Comment by munen** [comment]
> The difference in the browser is IMO a minor problem which can be ignored (and is not our fault ;) ). As long as the app doesn't crash but shows a error message or red border around the search input I'm fine :)

image

It will put the red border around search for anything greater or equal to `clock:2020-02-32`, though.

File: src/lib/headline_filter.js
Comment (munen) [reply src/lib/headline_filter.js:207]: "Exclusion" always trumps inclusion when searching. Think of it this way: When searching without a term, the default is to show everything. As soon as you introduce a constraint, it will remove terms from the resultset. They cannot be brought back by further inclusions.

Not really a search engine, but the behavior is the same:

```
$ cat > great_editors
emacs
vim
$ cat great_editors | grep -v vim | grep vim | wc -l
0
```

File: src/lib/headline_filter_parser.grammar.pegjs
Comment (munen) [reply src/lib/headline_filter_parser.grammar.pegjs]: Cool, thanks! If it helps, you're quite consistent in naming it `timestamp`:

```
munen@lambda:~/src/200ok/organice% grep -r timestamp src | wc -l
388
munen@lambda:~/src/200ok/organice% grep -r timeStamp src | wc -l
0
munen@lambda:~/src/200ok/organice% grep -r TimeStamp src | wc -l
0
```

**Comment by munen** [comment]
Update on the `date:` question

> Do you have a reason that you want date to include inactive timestamps? If we should have a search on inactive timestamps, I propose to add a dedicated flag for that like inactive: true.

When testing, I realized that `date:` does not include inactive timestamps, only active timestamps. Hence, the behavior is already as expected! :+1:

**Comment by tarnung** [comment]
> Update on the `date:` question
> 
> > Do you have a reason that you want date to include inactive timestamps? If we should have a search on inactive timestamps, I propose to add a dedicated flag for that like inactive: true.

I don't have a reason, just the general hunch that being able to search for something might generally be useful ;)
So we can focus on active timestamps for now and rethink should someone express a desire for that feature.

> When testing, I realized that `date:` does not include inactive timestamps, only active timestamps. Hence, the behavior is already as expected! 👍

That's not quite correct. at least I can find inactive deadlines and schedules. No inactive timestamps in header or description though. So the current behaviour is inconsistent. The easy fix is filtering for Timestamps being active.

**Comment by munen** [comment]
> I don't have a reason, just the general hunch that being able to search for something might generally be useful ;)
So we can focus on active timestamps for now and rethink should someone express a desire for that feature.

:+1: 

> That's not quite correct. at least I can find inactive deadlines and schedules. No inactive timestamps in header or description though. So the current behaviour is inconsistent. The easy fix is filtering for Timestamps being active.

I've only tested with inactive timestamps in header and description. Inctive timestamps in planning items sounds like an oxymoron to me(; Hence, I didn't test that. Technically it can be done, but the rationale of a planning item is to have it show up in the agenda whereas an inactive timestamp is to not have it show up. 

Emacs Org mode agrees wih that sentiment:

image
 
As does the organice agenda:

image

**Comment by tarnung** [comment]
> > * When I began to write tests, all the existing tests passed. Soon
> >   thereafter some tests randomly started to fail. They had nothing
> >   to do with what i did so I ignored them.
> 
> That doesn't sound alright. Can you point me to the tests that randomly started to fail? So far, one user reported that one test in particular was failing randomly. Personally I have seen only one test fail randomly once (on CI). Here's the issue we have created for it: #442
> 
> I haven't seen this behavior since. But if you're experiencing it, too, then it would certainly be good to track it down.

It's a whole bunch of tests. It seems to be a line break thing probably causes by silly me running windows.

>Test Suites: 5 failed, 5 passed, 10 total
>
>Tests:       37 failed, 328 passed, 365 total
>
>Snapshots:   1 failed, 2 passed, 3 total

**Comment by munen** [comment]
> It's a whole bunch of tests. It seems to be a line break thing probably causes by silly me running windows.

That could be it. organice has many places where it uses `\n` as a newline. Can you try using `\r\n` in one of the tests and see if it runs through? If so, we could replace all `\n` with `require('os').EOL;` (https://nodejs.org/api/os.html#os_os_eol).

**Comment by schoettl** [comment]
> If so, we could replace all `\n` with `require('os').EOL;` (https://nodejs.org/api/os.html#os_os_eol).

Oh that would affect probably thousands line of code, many of them must be converted from `"string\n"` to
```
import {EOL} from 'os';
`string${EOL}`
```

And what's about sample files that are loaded for tests?

I hope that we don't have to go that way and can just always use UNIX EOL `\n` internally.

**Comment by munen** [comment]
What's the alternative? Adding documentation that developing on Windows is not supported?(;

File: src/lib/headline_filter_parser.grammar.pegjs
Comment (schoettl) [reply src/lib/headline_filter_parser.grammar.pegjs:106]: Ah, pegjs doesn't support the "unordered alternative (|)" – but you could remove duplication when you replace `"|"` in my example above with `" / "`.

That would save us duplication of 9 tokens :)

**Comment by schoettl** [comment]
I guess I didn't really understand the problem of the failing tests. Is it because git on windows check out CRLF and therefore tests with strings like
```
`multi
line`
```
fail?

**Comment by tarnung** [comment]
I can confirm that changing the line endings of the loaded test org files to LF makes the errors disappear. 

> I guess I didn't really understand the problem of the failing tests. Is it because git on windows check out CRLF

Exactly. Maybe it could be fixed by me locally by adjusting git settings.

**Comment by schoettl** [comment]
It should be `core.autocrlf`

https://git-scm.com/book/it/v2/Customizing-Git-Git-Configuration

**Comment by munen** [comment]
If `git config core.autocrlf false` fixes the issues, that would definitively be preferable to changing all occurences of `\n`(;

That's also reasonable to put into the docs.

**Comment by schoettl** [comment]
... plus the editor/IDE must save LF, not CRLF.

**Comment by munen** [comment]
Maybe that can be addressed with `*       text=auto` (https://mirrors.edge.kernel.org/pub/software/scm/git/docs/gitattributes.html#_end_of_line_conversion).

**Comment by schoettl** [comment]
> @schoettl I think the requested changes from you are also addressed. If so, please 'accept' the changes.

The only thing that I think should be changed is [this](https://github.com/200ok-ch/organice/pull/505#discussion_r506886838).

@tarnung Great work, it's a really cool feature!

**Review by schoettl — APPROVED** [review]

**Comment by munen** [comment]
> The only thing that I think should be changed is this.

Sweet, I'll wait for @tarnung for feedback on that.

The last thing to add is documentation (to sample.org) and attribution before merging, then. I'll take care of those when ready.

File: src/lib/headline_filter_parser.grammar.pegjs
Comment (tarnung) [reply src/lib/headline_filter_parser.grammar.pegjs]: Oh, neat. I changed it.

Comment (schoettl) [reply src/lib/headline_filter_parser.grammar.pegjs]: Thanks!

**Comment by munen** [comment]
I've added some documentation. Do you think it's appropriate, is something missing?

File: sample.org
Comment (schoettl) [inline sample.org:338]: Sorry if I'm wrong, but I had in mind that `w` was "from beginning of this week" (and "to end of this week").

Comment (tarnung) [reply sample.org:338]: As it currently is implemented `date:1w` is equivalent to `date:..1w` or in words 'from now to one week in the future'.
`date:..w` is 'from now until the end of the week'
`date:w` is the current week (start to finish).

Comment (munen) [reply sample.org]: Thank you for your feedback! I have changed the documentation accordingly: https://github.com/200ok-ch/organice/pull/505/commits/4216e1655776143ec23cb36b9926a28b124d2d0d

I've also added information on negating filters.

**Comment by munen** [comment]
@tarnung I might misunderstand something in how search behaves or I might have found a bug. Can you take a look and comment on that, please?

image

**Comment by tarnung** [comment]
In case you searched today, you may have been fooled by the fact that a week is Sunday to Saturday.
I don't like it this way since I always considered Monday to be the start of the week, but the agenda view has Sunday as the first day of the week and it's also the default behaviour in the date-fn library, so I ran with it.

**Comment by munen** [comment]
@tarnung You are absolutely right. It’s also the default in Org mode. I forgot, because in Emacs this default can be changed. organice doesn’t support such a flag, yet, and I usually use the daily agenda(;

Thank you for checking it out!🙏🏻

**Comment by munen** [comment]
I have updated the docs that a "week" in organice is described as in the ["Western Traditional" system](https://en.wikipedia.org/wiki/Week#The_ISO_week_date_system).

File: sample.org
Comment (munen) [reply sample.org]: It's only been two days since I've asked for feedback, but I assume you're good with the changes, because both of you have been active in other issues/prs since(;

I'll go ahead, add the attribution to the changelog and will merge to `master`. If you do find something that want changed in the documentation, we can change it anytime!

**Review by munen — APPROVED** [review]

**Pull request merged**
